mfs/ops.go

```diff
@@ -12,60 +13,207 @@ import (
 	if err != nil {
 		return err
 	}
 
-	srcDirObj, err := lookupDir(r, srcDir)
+	dstNode, err := DirLookup(r.GetDirectory(), dst)
 	if err != nil {
+		if err == os.ErrNotExist {
+			return handleDstFileUnExisted(r, src, dst)
+		}
 		return err
 	}
 
-	srcObj, err := srcDirObj.Child(srcFname)
+	if dstNode.Type() == TDir {
+		dstDir, _ := dstNode.(*Directory)
+		if n, err := dstDir.Child(srcDirName); err == nil { // contains srcDir
+			empty, err := isEmptyNode(n)
+			if err != nil {
+				return err
+			}
+			if empty {
+				if err = dstDir.Unlink(srcDirName); err != nil {
+					return err
+				}
+			} else {
+				return ErrDirExists
+			}
+		}
+
+		if err = dstDir.AddChild(srcDirName, srcNode); err != nil {
+			return err
+		}
+		return srcParDir.Unlink(srcDirName)
+	} else if dstNode.Type() == TFile {
+		if dst[len(dst)-1] == '/' {
+			return errInvalidDirPath
+		}
+		return errMvDirToFile
+	} else {
+		return errInvalidNodeType
+	}
+}
+
+// moveFile moves the file at 'src' to 'dst'
+func moveFile(r *Root, src, dst string) error {
+	srcFileName, srcNode, srcParDir, err := getNodeAndParent(r, src)
 	if err != nil {
 		return err
 	}
 
-	nd, err := srcObj.GetNode()
+	dstNode, err := DirLookup(r.GetDirectory(), dst)
 	if err != nil {
+		if err == os.ErrNotExist {
+			return handleDstFileUnExisted(r, src, dst)
+		}
 		return err
 	}
 
-	fsn, err := dstDir.Child(filename)
-	if err == nil {
-		switch n := fsn.(type) {
-		case *File:
-			_ = dstDir.Unlink(filename)
-		case *Directory:
-			dstDir = n
-		default:
-			return fmt.Errorf("unexpected type at path: %s", dst)
+	if dstNode.Type() == TFile {
+		if dst[len(dst)-1] == '/' {
+			return errInvalidDirPath
+		}
+		// replace(src, dst)
+		dstFileName, _, dstParDir, err := getNodeAndParent(r, dst)
+		if err != nil {
+			return err
+		}
+
+		err = dstParDir.Unlink(dstFileName)
+		if err != nil {
+			return err
+		}
+
+		err = dstParDir.AddChild(srcFileName, srcNode)
+		if err != nil {
+			return err
 		}
-	} else if err != os.ErrNotExist {
+
+		return srcParDir.Unlink(srcFileName)
+	} else if dstNode.Type() == TDir {
+		dstDir, _ := dstNode.(*Directory)
+		if n, err := dstDir.Child(srcFileName); err == nil {
+			// contains child with srcFileName
+			empty, err := isEmptyNode(n)
+			if err != nil {
+				return err
+			}
+
+			if !empty && n.Type() == TDir {
+				return ErrDirExists
+			}
+
+			if err = dstDir.Unlink(srcFileName); err != nil {
+				return err
+			}
+		}
+
+		err := dstDir.AddChild(srcFileName, srcNode)
+		if err != nil {
+			return err
+		}
+
+		return srcParDir.Unlink(srcFileName)
+	} else {
+		return errInvalidNodeType
+	}
+}
+
+func handleDstFileUnExisted(r *Root, src, dst string) error {
+	err := Mkdir(r, dst, MkdirOpts{Mkparents: true, Flush: true})
+	if err != nil {
+		return errUnableCreateNonExistedFile
+	}
+	dstFileName, _, dstParDir, err := getNodeAndParent(r, dst)
+	if err != nil {
+		return err
+	}
+	if err := dstParDir.Unlink(dstFileName); err != nil {
```

Comment:
You're creating a directory and then deleting it?

Reply:
Create empty target and then replace it